From 99263abe1a6b1b2ca1cf88dd573c392a0154d843 Mon Sep 17 00:00:00 2001
From: arathinx <ayyappanx.rathinam@intel.com>
Date: Tue, 1 Dec 2020 14:59:41 +0800
Subject: [PATCH] UGW_SW-54036: sleeping function called from invalid context
 at kernel CQM driver is using spin lock irqsave before ltq_request_dma() API
 which has mutex_lock in it.

---
 drivers/net/ethernet/lantiq/cqm/grx500/cbm.c | 15 +++++----------
 1 file changed, 5 insertions(+), 10 deletions(-)

diff --git a/drivers/net/ethernet/lantiq/cqm/grx500/cbm.c b/drivers/net/ethernet/lantiq/cqm/grx500/cbm.c
index 6c9316f2fd40..f307fc46e4c2 100644
--- a/drivers/net/ethernet/lantiq/cqm/grx500/cbm.c
+++ b/drivers/net/ethernet/lantiq/cqm/grx500/cbm.c
@@ -1182,6 +1182,7 @@ static void cbm_qidt_set(const struct cbm_qidt_elm *qid_set,
 	struct cbm_qidt_elm qidt_elm;
 	int cls_done = 0, ep_done = 0, fl_done = 0, fh_done = 0,
 	dec_done = 0, enc_done = 0, mpe1_done = 0, mpe2_done = 0;
+	unsigned long sys_flag;
 
 	pr_debug("%d\n", qid_val);
 	pr_debug("%u %u %u %u %u %u %u %u\n", qid_set->clsid,
@@ -1200,7 +1201,8 @@ static void cbm_qidt_set(const struct cbm_qidt_elm *qid_set,
 		 qid_mask->dec_mask,
 		 qid_mask->flowid_hmask,
 		 qid_mask->flowid_lmask);
-#if 1
+
+	spin_lock_irqsave(&cbm_qidt_lock, sys_flag);
 	state = STATE_CLASS;
 	while (state != STATE_NONE) {
 		switch (state) {
@@ -1353,7 +1355,7 @@ static void cbm_qidt_set(const struct cbm_qidt_elm *qid_set,
 		break;
 		};
 		}
-#endif
+	spin_unlock_irqrestore(&cbm_qidt_lock, sys_flag);
 }
 
 static s32
@@ -1428,7 +1430,6 @@ cbm_queue_map_entry_t *entry,
 u32 flags
 )
 {
-	unsigned long sys_flag;
 	struct cbm_qidt_elm qidt_elm;
 	struct cbm_qidt_mask qidt_mask = {0};
 
@@ -1461,9 +1462,7 @@ u32 flags
 	if (flags & CBM_QUEUE_MAP_F_TC_DONTCARE)
 		qidt_mask.classid_mask = 1;
 
-	spin_lock_irqsave(&cbm_qidt_lock, sys_flag);
 	cbm_qidt_set(&qidt_elm, &qidt_mask, tmu_queue_id);
-	spin_unlock_irqrestore(&cbm_qidt_lock, sys_flag);
 	return CBM_SUCCESS;
 }
 
@@ -3370,7 +3369,6 @@ dp_enable(
 	u32 num;
 	cbm_tmu_res_t *res = NULL;
 	int q_buff_num;
-	unsigned long sys_flag;
 	int port, port_high;
 	struct tmu_equeue_drop_params thx = {0};
 
@@ -3411,7 +3409,7 @@ dp_enable(
 						      0);
 		}
 	}
-	spin_lock_irqsave(&cbm_qidt_lock, sys_flag);
+
 	if (local_entry && ((local_entry->egp_type == DP_F_DIRECT) ||
 			    (local_entry->egp_type == DP_F_FAST_DSL) ||
 			    (local_entry->egp_type == DP_F_DIRECTLINK))) {
@@ -3467,7 +3465,6 @@ dp_enable(
 	tmu_egress_port_enable(res[0].tmu_port, (flags & CBM_PORT_F_DISABLE) ? 0 : 1);
 UNLOCK:
 	kfree(res);
-	spin_unlock_irqrestore(&cbm_qidt_lock, sys_flag);
 	return CBM_SUCCESS;
 }
 
@@ -3928,11 +3925,9 @@ CBM_F_QUEUE_DROP flag from the DDR table*/
 		CBM_F_QUEUE_DROP flag. For each such entry, program the CBM Queue Map table entry in CBM to map
 		to the Global Drop Queue*/
 		if (flags & CBM_Q_F_DISABLE) {
-			spin_lock_irqsave(&cbm_qidt_lock, sys_flag);
 			cbm_qidt_set(&qidt_set,
 				     &qidt_mask,
 				     255);
-			spin_unlock_irqrestore(&cbm_qidt_lock, sys_flag);
 		}
 		/* For each queue_id in the CBM Queue map table matched entries
 		 *(by EP), disable/enable the TMU queue by calling
